# Quantum Bit Shifting (Right and Left) Using Qiskit
This implementation demonstrates how to shift the bits of a 4-qubit quantum register to the **right** and **left** using Qiskit. 

We will:
1. Initialize a quantum circuit with the input bits.
2. Apply swap operations to perform the desired shift.
3. Measure the qubits to observe the shifted state.


In [1]:
from qiskit import QuantumCircuit, transpile
from qiskit_aer import AerSimulator

## Step 1: Import Necessary Libraries
- **Qiskit**: The framework for quantum programming.
  - QuantumCircuit: For building the quantum circuit.
  - transpile: To optimize the circuit for the simulator.
- **AerSimulator**: A quantum simulator backend for running and testing circuits.


In [ ]:
def create_shift_circuit(input_bits, direction):
    qc = QuantumCircuit(4)

    # Initialize the input bits
    for i, bit in enumerate(reversed(input_bits)):
        if bit == '1':
            qc.x(i)


    if direction == "right":
        qc.swap(0, 1)
        qc.swap(1, 2)
        qc.swap(2, 3)
    elif direction == "left":
        qc.swap(2, 3)
        qc.swap(1, 2)
        qc.swap(0, 1)
    else:
        raise ValueError("Direction must be 'right' or 'left'.")


    qc.measure_all()

    # Simulate the circuit
    simulator = AerSimulator()
    compiled_circuit = transpile(qc, simulator)
    job = simulator.run(compiled_circuit)
    result = job.result()

    return result.get_counts()


## Step 2: Define the create_shift_circuit Function
This function creates a quantum circuit and performs a shift operation. It has the following steps:

### Input Parameters
1. **input_bits**: A binary string (e.g., "1011") representing the initial state of the qubits.
2. **direction**: Specifies whether to shift **right** or **left**.

### Circuit Construction
1. **Initialization**:
   - Qubits are initialized to the state represented by the binary string.
   - For each bit that is 1, the corresponding qubit is flipped to \( |1\rangle \) using the x gate.

2. **Swap Operations**:
   - Swaps are applied between adjacent qubits based on the desired direction:
     - **Right Shift**: \( Q_0 \leftrightarrow Q_1 \), \( Q_1 \leftrightarrow Q_2 \), \( Q_2 \leftrightarrow Q_3 \).
     - **Left Shift**: \( Q_2 \leftrightarrow Q_3 \), \( Q_1 \leftrightarrow Q_2 \), \( Q_0 \leftrightarrow Q_1 \).

3. **Measurement**:
   - All qubits are measured to record the final state.

4. **Simulation**:
   - The circuit is simulated using AerSimulator to obtain the measurement results.

### Output
The function returns a dictionary of measurement results (counts), showing how many times each final state occurred.


In [3]:
# Input binary string
input_bits = "1011"

# Perform right shift
right_shift_result = create_shift_circuit(input_bits, "right")

# Perform left shift
left_shift_result = create_shift_circuit(input_bits, "left")


## Step 3: Define Input and Perform Shifts
1. **Input Binary String**:
   - input_bits: The initial state of the qubits, represented as a binary string (e.g., "1011").
   
2. **Perform Right Shift**:
   - Call create_shift_circuit(input_bits, "right") to shift the bits to the right.
   
3. **Perform Left Shift**:
   - Call create_shift_circuit(input_bits, "left") to shift the bits to the left.


In [4]:
# Display the results
print('Right shift:\t' + max(right_shift_result))
print('Left shift:\t' + max(left_shift_result))


Right shift:	1101
Left shift:	0111


## Step 4: Display Results
- The results of the simulation are displayed as binary strings:
  - **Right Shift**: The output after shifting the bits to the right.
  - **Left Shift**: The output after shifting the bits to the left.
- The max() function retrieves the most frequent measurement result, as we expect a single deterministic output in this quantum operation.

### Example Output
For input_bits = "1011":
- **Right Shift**: 0101
- **Left Shift**: 0111
